include/fiction/algorithms/simulation/sidb/operational_domain.hpp

```diff
@@ -0,0 +1,1266 @@
+                      [this](const auto& sp) { is_operational(sp); });
+
+        // a queue of (x, y) dimension step points to be evaluated
+        using parameter_queue = std::queue<step_point>;
+        parameter_queue queue{};
+
+        // a utility function that adds the adjacent points to the queue for further evaluation
+        const auto queue_next_points = [this, &queue](const step_point& sp)
+        {
+            for (const auto& m : moore_neighborhood(sp))
+            {
+                if (!has_already_been_sampled(m))
+                {
+                    queue.push(m);
+                }
+            }
+        };
+
+        // add the neighbors of each operational point to the queue
+        for (const auto& [param_point, status] : op_domain.operational_values)
+        {
+            if (status == operational_domain::operational_status::OPERATIONAL)
+            {
+                queue_next_points(to_step_point(param_point));
+            }
+        }
+
+        // for each point in the queue
+        while (!queue.empty())
+        {
+            // fetch the step point and remove it from the queue
+            const auto sp = queue.front();
+            queue.pop();
+
+            // if the point has already been sampled, continue with the next
+            if (has_already_been_sampled(sp))
+            {
+                continue;
+            }
+
+            // check if the point is operational
+            const auto operational_status = is_operational(sp);
+
+            // if the point is operational, add its four neighbors to the queue
+            if (operational_status == operational_domain::operational_status::OPERATIONAL)
+            {
+                queue_next_points(sp);
+            }
+        }
+
+        log_stats();
+
+        return op_domain;
+    }
+    /**
+     * Performs contour tracing to determine the operational domain. The algorithm first performs a random sampling of
+     * up to the specified number of samples. It stops random sampling once it finds a single operational point, from
+     * which it moves straight outwards until it encounters the counter of the operational domain. From this point, it
+     * traces the contour until it reaches the initial contour point again. The operational domain will finally only
+     * contain up to `samples` random non-operational points as well as the contour of the found operational domain plus
+     * a one pixel wide border around it.
+     *
+     * @param samples Maximum number of random samples to be taken before contour tracing.
+     * @return The (partial) operational domain of the layout.
+     */
+    [[nodiscard]] operational_domain contour_tracing(const std::size_t samples) noexcept
+    {
+        mockturtle::stopwatch stop{stats.time_total};
+
+        // first, perform random sampling to find an operational starting point
+        const auto starting_point = find_operational_step_point_via_random_sampling(samples);
```

Comment:
Yes, there is. For now, contour tracing determines a single starting point and works from there. As such, it can abort random sampling as soon as such a starting point is found. Flood fill on the other hand, always performs the number of given random samples in order to potentially find operational points in disconnected domains to expand from. Contour tracing does not yet work with disconnected domains.

Reply:
Makes sense. Thanks.